### ISSUE TITLE
Buildings overlay failing to make many buildings editable

### ISSUE THREAD

**Issue by davidpnewton** [streetcomplete/StreetComplete#5506]
There are considerable numbers of buildings that the new Buildings overlay us not picking up as such and thus not allowing the editing of.

Look at the buildings in the French commune of St Etienne au Mont just to the south east of Boulogne-sur-Mer to see examples of this.

**How to Reproduce**

Compare way 132262690 with way 132264085 for an example of two buildings directly next to each other. Difference appears due to something with building levels or building roof type tagging.

**Expected Behavior**
Both way 132262690 and way 132264085 picked up as  buildings by the overlay and thus editable.

**Versions affected**
Android 14
Streetcomplete 57.0 beta-2

**Comment by matkoniecz** [comment]
https://www.openstreetmap.org/way/132262690

https://www.openstreetmap.org/way/132264085

buildings with  `building:use` keys currently are skipped:

https://github.com/streetcomplete/StreetComplete/blob/2aa9bb6ccc3fc974636581166fc935e76404de4b/app/src/main/java/de/westnordost/streetcomplete/overlays/buildings/BuildingsOverlay.kt#L51

**Comment by mnalis** [comment]
Makes sense to me. Unless quest is going to be extended to handle such cases (with special subquestions, confirmations, or whatever), this should probably be added to [SC FAQ](https://wiki.openstreetmap.org/wiki/StreetComplete/FAQ#Overlays) as it sounds confusing.

**Comment by westnordost** [comment]
It is pretty much a missing feature to reduce implementation complexity of that overlay.

If this is implemented (by someone), "historic", "ruins", "disused" (etc?) should be handled accordingly. E.g.

`building=industrial + building:use=apartments` -> "Built as industrial building, now it is apartment building"
`building=industrial + ruins=yes` -> "Built as industrial building, now it is ruins"

**Comment by call-me-matt** [comment]
I understand and agree to ignore complex buildings. But for a better user experience, I suggest to give it a (greyed out) color code. As it is now, the user thinks the building tag is not correct or the map data not up to date. This triggers experimenting with the data which is contra-productive.

<table border=0><tr><td>
as-is:</td><td>to-be:</td></tr><tr><td>
<img src=https://github.com/streetcomplete/StreetComplete/assets/34400929/a688af7c-93a8-4bd7-b460-2a752e4d31c7 width=250px />
</td><td>
<img src=https://github.com/streetcomplete/StreetComplete/assets/34400929/bf3c0e21-d321-46fa-a688-9bc901903696 width=250px />
</td></tr></table>

**Comment by westnordost** [comment]
Experimenting? You cannot, as you cannot click on it, or can you?

**Comment by call-me-matt** [comment]
Well, you are right. But when buildings do not show up in StreetComplete, people use other editors to "fix" the building that seems to be not correctly mapped. I have seen it at different locations. I was also confused and searched here before understanding the logic behind. In my point of view a color code can improve the user experience here.

**Comment by Amaanprobably** [comment]
Hi @westnordost, I saw you'd held off on this until the Compose migration was done. Now that #5616 is closed, is the Buildings overlay work here unblocked? If so, I'd like to pick this up: extending the filter to handle building:use, and generating appropriate messaging for combinations like building:use + historic/ruins/disused per your comment above. Let me know if there's a preferred approach or anything I should watch out for before I start.

**Comment by westnordost** [comment]
It's unblocked, yeah. Sure, go ahead! When you create a PR, please disclose if and if yes in what way you used the help of an agent and always write the PR description yourself.

---

### UI

UI-wise, I'd expect the normal UI to not change. Only when clicking on a new item in the …-menu labeled like "Different use than constructed as" (or similar/shorter) - or if the selected building already has `building:use` etc tagged, obviously -,  the UI switches to showing two of these building type selects nexts to each other. Much like the form for the surface overlay shown on `highway=footway + bicycle=designated + segregated=yes`.

In that layout, the quick-select to the bottom would not be shown. and each side would be labelled so that it is understood easily. E.g. maybe
- constructed as
- …but now it's

Probably, there should be another item in the …-menu to go back to the normal layout ("Used as constructed as"?). Or that would be done automatically when one selects the same building type in both sides?

There is a whole area of edge cases caused by `ruins=yes`, `disused=yes` etc. though, that not only have to be taken into account in the UI but first and foremost also in the tagging and parsing logic.

### Tagging and parsing logic

Anyway, besides the layout, both the filter needs to be adapted to display and allow editing of such buildings. Also, tagging logic needs to be adapted and tested in unit tests.

I am a bit unsure about how to deal with `ruins=yes`, `disused=yes` and in particular `historic=yes`.
"Constructed as… [school building] …but now it's [disused building]" and the same with ruins seems a good fit, but I expect that there are quite a few edge cases to take into account. Also with parsing (and then, in the end, tagging the user's selection). What if we encounter, in the wild, buildings that are tagged with both `ruins=yes` but also with `building:use` and `building`.

Plus, `historic=yes` would require additional handling "Constructed as … [office building] … but now it's [historic building]" is awkward - the reason why StreetComplete is fine with `building=yes + historic=yes` → "historic building" is because for historic buildings, it is usually quite difficult to say what exactly was their original use and what they are used for now. In any case, maybe `historic` would need to be an additional checkbox or something. Maybe for now, don't change the treatment of historic buildings unless you have a good idea for it.

### Adapt what is selectable

Question is really, what should be displayed for a building that has `building=yes + disused=yes` tagged? After all, especially for ruins, it may be impossible to know what the building was originally constructed as.

Side-by-side layout with

1. "Constructed as… [select…] …but now it's [disused building]"

2. or single layout with [disused building]

…? In any case, when the former layout is currently shown, it wouldn't make sense when "Disused building" was selectable on the left side.

### Adapt rendering of buildings

Without even an addition to `OverlayStyle`, buildings that have both `building:use` and and a different `building` tagged, could be shown differently in the highlighting. I'd propose that if `building:use` is tagged, color the fill area of the building like that, and then also add an outline with the `building`s color.

**Comment by matkoniecz** [comment]
> What if we encounter, in the wild, buildings that are tagged with both ruins=yes but also with building:use and building.

and this is valid at least for ruins now used for tourism purposes, for example ruined castle building

<img width="1200" height="800" alt="Image" src="https://github.com/user-attachments/assets/6b681631-37bd-4214-8301-8d9e7ae392ad" />

https://polskapogodzinach.pl/jura-krakowsko-czestochowska-atrakcje/

**Comment by westnordost** [comment]
Not relevant. Not tagged as a building and not with `building:use`. Only with both `ruins=yes` and `historic=yes`.

https://www.openstreetmap.org/way/14149043#map=17/50.453047/19.552118

**Comment by Amaanprobably** [comment]
> ### Adapt rendering of buildings
> Without even an addition to `OverlayStyle`, buildings that have both `building:use` and and a different `building` tagged, could be shown differently in the highlighting. I'd propose that if `building:use` is tagged, color the fill area of the building like that, and then also add an outline with the `building`s color.

Was navigating through the codebase to implement a different outline for `building:use` but failed to see how that was possible without changing `OverlayStyle.Polygon`, as the existing outline color is just a `color.darkened()` version of the fill color (`building`'s own color) :

```kotlin
p.addProperty("outline-color", overlayStyle.color.darkened().toRgbaString())
```

I was thinking of adding an optional `outline: Color? = null` field to `OverlayStyle.Polygon` (as it defaults to null, existing behaviour remains unaffected) if that is okay with you:

```kotlin
data class Polygon(
    val color: Color,
    //New addition
    val outline: Color? = null,
    val icon: DrawableResource? = null,
    val label: String? = null,
    val height: Float? = null,
    val minHeight: Float? = null,
    override val disabled: Boolean = false,
) : OverlayStyle
```
That field would then be wired into the rendering like this, falling back to today's exact behavior when `outline` isn't set:

```kotlin
// before
p.addProperty("outline-color", overlayStyle.color.darkened().toRgbaString())

// after
p.addProperty("outline-color", overlayStyle.outline?.toRgbaString() ?: overlayStyle.color.darkened().toRgbaString())
```

Let me know if you had a different approach in mind.

**Comment by westnordost** [comment]
> Was navigating through the codebase to implement a different outline for building:use but failed to see how that was possible 

For one building element, you could just emit two pairs of `element to Polygon(...)` + `element to Polyline(...)` and then use `flatMap` instead of `map`.

<strike>But your solution is also fine. It should be `val outline: Stroke? = null,`, though.</strike>

**Edit**: But I like your solution better.
